### ISSUE TITLE
smoothscroll

### ISSUE THREAD

**Issue by satazor** [taion/scroll-behavior#112]
Hi @taion 

I would like to apply smooth scrolling when transitioning between pages. At the moment you are calling `window.scrollTo(x, y)` but I would like to use `window.scrollTo({ top: y, left: x, behavior: 'smooth' })` which has been implemented in Firefox and there's a [polyfill](https://github.com/iamdustan/smoothscroll) for other browsers.

I was thinking if we could create an option, e.g.: `scrollTo` which defaults to `window.scrollTo`, to allow customization of this behavior.

Let me know what you think. Thanks in advance.

**Comment by nelsongustavo** [comment]
+1

**Comment by taion** [comment]
Are you still looking at this? It would make sense to pass in a callback to handle the actual scrolling.

**Comment by vasco-santos** [comment]
Hello,

We need to define the options of scroll for a project. Therefore, we had a brainstorm, in order to identify the possible solutions for integrating the feature in this library.

**1. Add a scrollToTarget option**

Essentially, when this option is specified it would be called instead of the internal _scrollToTarget function.

Additionally, besides the `element` and `target`, it should be called with an additional third argument which is the location object. This allows deciding the scroll options based on the location, for instance (assuming an array is received):

```
scrollToTarget = (element, target, location) => {

	const [x, y] = target;

	window.scrollTo(x, y, { …location.state.scrollOptions });
}
```

The downside of this approach consists of being necessary to replicate some logic that is already in the library.

**2. Callback returning an Array with 3 entries**

We could return an array with an additional entry for defining the scroll options:

`[X, Y, scrollOptions]`

We do not appreciate this solution for the following reasons:

- Not allowing scroll options customization on scroll to element;
- When it is not intended to scroll to a particular (X, Y), it would be strange to define X and Y as `undefined`. Consequently, this module would be responsible for scrolling to the saved position if the coordinates were `undefined`.

**3. Callback returning an Object and refactor of the other return options**

We could return an object for customizing the scroll, which would uniform the array and element option into a unique option, as well as allowing a scroll options definition.

```
{
	x: _ ,
	y: _ ,
	scrollOptions: _ ,
}
```

```
{
	element: _ ,
	scrollOptions : _ ,
}
```

Regarding this solution, we don’t like the following points:

- In the case of the lib consumer had set x, y and element, we would have to decide if we should scroll to the element, or to the position;
- When it is not intended to scroll to a particular (X, Y) or element, it would be strange to define only the scrollOptions. In addition, this module would be responsible for scrolling to the saved position;
- This solution would cause a breaking change for the lib (unless we keep all other redundant options).

**4. Callback returning a callback for controlling the scroll options**

We could return a callback, which would be called by the lib once the **_scrollToTarget** is called. This callback would receive **(element, target)** and the lib consumer would scroll as he wishes to the target position (saved position).

```
shouldUpdateScroll: (prevRouterState, routerState) => updateScrollCallback
```


```
function updateScrollCallback(element, target) {

	const [x, y] = target;

	window.scrollTo(x, y, { behavior: 'smooth' });
}
```

In the context of this solution, we consider that the callback returning another callback may be strange for the consumers, as well as more complex. However, it will provide more freedom for the consumer.

What do you think @taion ?

**Comment by taion** [comment]
What do you think of doing something like (1), except:

1. Subclass `ScrollBehavior` and override `scrollToTarget`
2. In the integrations, we accept a factory method (`createScrollBehavior`) where you can use your own `ScrollBehavior` subclass

I'm a bit on the fence between that approach and accepting the callback for handling the scrolling. I feel like part of the benefit of using a class for `ScrollBehavior` is exactly so users can inherit and override methods for things like actually doing the scroll, but maybe it's less clunky to just accept the callback.

Interested to hear your thoughts on this "subclass-and-replace-factory-method" v. your (1).

**Comment by vasco-santos** [comment]
Hello @taion 

Regarding your proposal, I like your idea behind the `subclass-and-replace-factory-method` suggestion. However, that approach will force the `react-router-scroll` to be updated, in order to allow this option/factory.

Therefore, my suggestion would be to implement the callback suggestion by now, which has a really simple implementation to allow the customization of the scroll and it can be used directly by projects using `react-router-scroll`. Maybe later, you could open a new issue for implementing the `subclass-and-replace-factory-method` approach.

**Comment by taion** [comment]
Fortunately I control both libraries, so I'm free to update either or both, and you'd need to change the code on your side anyway.

**Comment by satazor** [comment]
@taion I think that subclassing is kind of weird, specially because we would be overriding private methods to achieve what I initially wanted. Also, it kind of increases the surface area for breaking changes, because basically any change in the base class might break projects.

Regarding the other proposals, what's the one that you must prefer? 1, 2, 3 or 4?

**Comment by taion** [comment]
The callback looks nicest, but of course I’d make the scroll method public.

**Comment by satazor** [comment]
Are you referring to the solution nr. **4**?

**Comment by taion** [comment]
@satazor What do you think of https://github.com/taion/scroll-behavior/pull/125 and https://github.com/taion/react-router-scroll/pull/86?

**Comment by taion** [comment]
Oops, I got confused because your avatars looked very similar. cc @vasco-santos as well.

**Comment by taion** [comment]
Released in scroll-behavior v0.9.5 + react-router-scroll v0.4.4.
